// mineflayer-cmd/lib/CommandManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandManager = void 0;
const Command_1 = require("./Command");
const Tokenizer_1 = require("./Tokenizer");
class CommandManager {
    constructor(bot) {
        this.commands = [];
        this.log = (sender, message) => {
            if (sender === '[CONSOLE]')
                console.log(message);
            else
                bot.chat(message);
        };
    }
    registerCommand(cmdName, handler, help = "", usage = "") {
        if (help === '' || usage === '') {
            console.warn(`[mineflayer-cmd] Note that leaving command description and usage information is not recommended. (Effected command: '${cmdName}')`);
        }
        const command = new Command_1.Command(cmdName, handler, help, usage);
        this.commands.push(command);
        return command;
    }
    setLogger(logger) {
        this.log = logger;
    }
    run(sender, command, cb = () => { }) {
        const tokens = Tokenizer_1.parseTokens(command);
        if (tokens.length === 0) {
            cb(new Error("Cannot parse empty string!"));
            return;
        }
        const cmd = this.commands.find(c => c.name === tokens[0]);
        if (!cmd) {
            cb(new Error("Command not found!"));
            return;
        }
        try {
            const { flags, args } = Tokenizer_1.extractElements(tokens, cmd);
            cmd.handler(sender, flags, args, cb);
        }
        catch (err) {
            cb(err);
        }
    }
}
exports.CommandManager = CommandManager;
